File: code/cryptography/baconian_cipher/baconian.php
<?php
/**
* Bacon's (aka Baconian) cipher (PHP)
* based on https://en.wikipedia.org/wiki/Bacon%27s_cipher
* implemented by Stefano Amorelli <amoste@outlook.it>
* Part of Cosmos by OpenGenus Foundation
*/

class BaconianCipher
{
  private $mapping;

  function __construct() {
    $this->mapping = [];
    $this->initMapping();
  }

  public function encrypt($plainText) {
    if (!ctype_alpha($plainText)) {
      throw new Exception('Plaintext should be alphabetical strings (no spaces, symbols and numbers).');
    }
    $plainText      = strtoupper($plainText);
    $encryptedText  = [];
    foreach (str_split($plainText) as $character) {
      array_push($encryptedText, $this->mapping[$character]);
    }
    return implode(' ', $encryptedText);
  }

  public function decrypt($encryptedText) {
    $plainText = [];
    foreach (explode(' ', $encryptedText) as $character) {
      foreach ($this->mapping as $key => $value) {
        if ($value === $character) {
          $plainCharacter = $key;
          continue;
        }
      }
      array_push($plainText, $plainCharacter);
    }
    return implode('', $plainText);
  }

  private function initMapping() {
    $this->mapping['A'] = 'aaaaa';
    $this->mapping['B'] = 'aaaab';
    $this->mapping['C'] = 'aaaba';
    $this->mapping['D'] = 'aaabb';
    $this->mapping['E'] = 'aabaa';
    $this->mapping['F'] = 'aabab';
    $this->mapping['G'] = 'aabba';
    $this->mapping['H'] = 'aabbb';
    $this->mapping['I'] = 'abaaa';
    $this->mapping['J'] = 'abaab';
    $this->mapping['K'] = 'ababa';
    $this->mapping['L'] = 'ababb';
    $this->mapping['M'] = 'abbaa';
    $this->mapping['N'] = 'abbab';
    $this->mapping['O'] = 'abbba';
    $this->mapping['P'] = 'abbbb';
    $this->mapping['Q'] = 'baaaa';
    $this->mapping['R'] = 'baaab';
    $this->mapping['S'] = 'baaba';
    $this->mapping['T'] = 'baabb';
    $this->mapping['U'] = 'babaa';
    $this->mapping['V'] = 'babab';
    $this->mapping['W'] = 'babba';
    $this->mapping['X'] = 'babbb';
    $this->mapping['Y'] = 'bbaaa';
    $this->mapping['Z'] = 'bbaab';
  }
}


/* Example usage: 

$cipher    = new BaconianCipher();
  
$encrypted = $cipher->encrypt('LOREMIPSUMDOLOR');

$decrypted = $cipher->decrypt($encrypted);

echo $decrypted; // output: LOREMIPSUMDOLOR

*/

File: code/cryptography/src/baconian_cipher/baconian.cpp
#include <map>
#include <string>

#define VALUE_LENGTH 5 // The length of a value in a key value pair for a baconian cipher

// Part of Cosmos by OpenGenus Foundation
// Typed out map
std::map<char, std::string> cipher = {
{ 'A', "AAAAA" }, { 'B', "AAAAB" }, { 'C', "AAABA" },
{ 'D', "AAABB" }, { 'E', "AABAA" }, { 'F', "AABAB" }, 
{ 'G', "AABBA" }, { 'H', "AABBB" }, { 'I', "ABAAA" }, 
{ 'J', "ABAAB" }, { 'K', "ABABA" }, { 'L', "ABABB" }, 
{ 'M', "ABBAA" }, { 'N', "ABBAB" }, { 'O', "ABBBA" }, 
{ 'P', "ABBBB" }, { 'Q', "BAAAA" }, { 'R', "BAAAB" }, 
{ 'S', "BAABA" }, { 'T', "BAABB" }, { 'U', "BABAA" }, 
{ 'V', "BABAB" }, { 'W', "BABBA" }, { 'X', "BABBB" }, 
{ 'Y', "BBAAA" }, { 'Z', "BBAAB" } 
};

// This function generates the exact map typed above
std::map<char, std::string> createBaconianMap()
{
    std::map<char, std::string> cipher;
    std::string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    std::string value = "";
    for (int i = 0; i < alphabet.length(); i++) 
    {
        for (int j = 0; j < VALUE_LENGTH; j++)
        {
	    if ((i & (1 << j)) == 0)
                value = 'A' + value;
            else
                value = 'B' + value;
        }
        char key = alphabet[i];
        cipher[key] = value;
        value = "";
    }
    return cipher;
}

std::string encrypt(std::map<char, std::string> &cipher, std::string message)
{
    std::string newMessage = "";
    for (int i = 0; i < message.length(); i++)
    {
        if (message[i] != ' ')
            newMessage.append(cipher[char(toupper(message[i]))]);
        else
            newMessage.append(" ");
    }
    return newMessage;
}

std::string decrypt(std::map <char, std::string> &cipher, std::string message)
{
    std::string newMessage = "";
    std::string currentString = "";
    int currentStringLength = 0;
    for (int i = 0; i < message.length() + 1; i++)
    {
        if (currentStringLength == VALUE_LENGTH)
        {
	    std::map<char, std::string>::iterator it = cipher.begin();
            while (it->second != currentString)
                it++;
            newMessage += it->first;
            currentString = "";
            currentStringLength = 0;
        }
        
        if (message[i] == ' ')
            newMessage.append(" ");   
        else
        {
            currentString += message[i];
            currentStringLength++;
        }
    }
    return newMessage;
}
